onmt/decoders/decoder.py

```diff
@@ -168,12 +169,28 @@ def _fix_enc_hidden(hidden):
             return hidden
 
         if isinstance(encoder_final, tuple):  # LSTM
-            return RNNDecoderState(self.hidden_size,
-                                   tuple([_fix_enc_hidden(enc_hid)
-                                          for enc_hid in encoder_final]))
+            self.state["hidden"] = tuple([_fix_enc_hidden(enc_hid)
+                                          for enc_hid in encoder_final])
         else:  # GRU
-            return RNNDecoderState(self.hidden_size,
-                                   _fix_enc_hidden(encoder_final))
+            self.state["hidden"] = (_fix_enc_hidden(encoder_final), )
+
+        # Init the input feed.
+        batch_size = self.state["hidden"][0].size(1)
+        h_size = (batch_size, self.hidden_size)
+        self.state["input_feed"] = \
+            self.state["hidden"][0].data.new(*h_size).zero_().unsqueeze(0)
+        self.state["coverage"] = None
+
+    def map_batch_fn(self, fn):
+        self.state["hidden"] = tuple(map(lambda x: fn(x, 1),
+                                         self.state["hidden"]))
+        self.state["input_feed"] = fn(self.state["input_feed"], 1)
+
+    def detach(self):
```

Comment:
I suggest naming this method `detach_state`.

Additionally, this method can be implemented using `map_batch_fn` (or `map_state`). This way, only the base class needs to define it (an abstract base class `Decoder` would be helpful here).

```python
self.map_state(lambda x, _: x.detach())
```

Reply:
Will do this later when creating an abstract base class.
would like to do some other renaming first.